File: python/valid_palindrome.py
class Solution(object):
    # O(n) time | O(1) space
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        i = 0
        j = len(s) - 1

        while i <= j:
            while i < j and not self.isAlphaNum(s[i]):
                i = i + 1
            
            while j > i and not self.isAlphaNum(s[j]):
                j = j - 1

            if s[i].lower() != s[j].lower():
                return False

            i = i + 1
            j = j - 1
        
        return True
    
    def isAlphaNum(self, c):
        return (ord('A') <= ord(c) <= ord('Z') or (ord('a') <= ord(c) <= ord('z') or (ord('0') <= ord(c) <= ord('9'))))
    
solution = Solution()

print(solution.isPalindrome("A man, a plan, a canal: Panama"))

File: src/main/java/june_2025/ValidPalindrome.java
package june_2025;

public class ValidPalindrome {

    public static void main(String[] args) {
        String s = "A man, a plan, a canal: Panama";

        ValidPalindrome validPalindrome = new ValidPalindrome();
        boolean result = validPalindrome.isPalindrome(s);
        System.out.println(result);
    }

    // O(n) time | O(1) space
    public boolean isPalindrome(String s) {
        int i = 0;
        int j = s.length() - 1;
        String lower = s.toLowerCase();

        while (i <= j) {
            while (i < j && !Character.isLetterOrDigit(lower.charAt(i))) {
                i++;
            }
            while (j > i && !Character.isLetterOrDigit(lower.charAt(j))) {
                j--;
            }
            if (lower.charAt(i) != lower.charAt(j)) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }

}
